django_spanner/compiler.py

```diff
@@ -38,76 +39,173 @@ def get_combinator_sql(self, combinator, all):
         :returns: A tuple containing SQL statement(s) with some additional
                   parameters.
         """
-        features = self.connection.features
-        compilers = [
-            query.get_compiler(self.using, self.connection)
-            for query in self.query.combined_queries
-            if not query.is_empty()
-        ]
-        if not features.supports_slicing_ordering_in_compound:
-            for query, compiler in zip(self.query.combined_queries, compilers):
-                if query.low_mark or query.high_mark:
-                    raise DatabaseError(
-                        "LIMIT/OFFSET not allowed in subqueries of compound "
-                        "statements."
-                    )
-                if compiler.get_order_by():
-                    raise DatabaseError(
-                        "ORDER BY not allowed in subqueries of compound "
-                        "statements."
-                    )
-        parts = ()
-        for compiler in compilers:
-            try:
-                # If the columns list is limited, then all combined queries
-                # must have the same columns list. Set the selects defined on
-                # the query on all combined queries, if not already set.
-                if (
-                    not compiler.query.values_select
-                    and self.query.values_select
+        if USING_DJANGO_3:
```

Comment:
This big if/else looks pretty bad:
1. they have a lot of common code that can be put in 1 place
2. the actual different logic between django 3/4 seems to be small but critical, does it make sense to provide a short comment, and minimize the length of code in this if/else

Reply:
We are decoupling the code for different Django versions so that if we need to make any changes for Django 4 it should be easy to do. Also I dont see it easy to have common code as there are significant differences in 3/4

Reply:
Could we add a comment to this method that explains __why__ we are overriding it, and highlight the line(s) that are actually different from the core Django implementation (so add an explicit comment where it is different)? That makes it easier to understand this, and easier the next time that a major version upgrade is being done.

Reply:
@ankiaga there are a few examples:
1. looks like L53 and L127 can be consolidated
2. seems that L43~L63 can be replaced by same code in L120~L144 as `compiler.query` is always there in both Django 3/4
3. seems that L64~L118 can be replaced with Django 4 logic too, there are only 2 real differences at L80(vs L162) and L105(vs L195)

I don't have my IDE setup to run unit tests, but by taking a glance, these real differences can be put in multiple smaller if/else blocks with the majority common code out of the big if/else

Reply:
@olavloite Added relevant comments in the code

@jaz-la Thanks for the explanation but as mentioned we don't want to optimize the code for different versions and keep it decoupled unless they are very straight forward .

Reply:
Repeating the code violates the [DRY principle](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), here likely >90% of the code in the if/else are the same, it has potential issues:
1.it's error prone and a maintenance burden, whatever change needed in the common code need to be made twice in the future
2. in the future if we need to support Django 5.2, this part will become a tech debt to clean up if django 5.2 has another minor change needed

As long as it works it's fine for me as a user, will defer it to other reviewers.

Reply:
The reason that we are reluctant to combine these two is exactly for maintenance reasons:
1. Django 3.2 will be deprecated, and is therefore extremely unlikely to receive any further updates. That means that this code block for Django 3.2 is also extremely unlikely to require any changes in the future.
2. It is conceivable that Django 4.2 will change things related to this code block, that might require us to change things here specifically for Django 4.2. Having a shared block with Django 3.2 would only complicate such changes.
3. For a future upgrade to Django 5.x, it is potentially a lot easier if we have a separate code block for Django 4.x, as that means that during such an upgrade we can consider whether having a shared block for Django 4.x and Django 5.x makes sense. Having a code block that is shared between Django 3.2 and 4.2 would make further sharing with Django 5.x even harder.